apps/aehttp/src/aehttp_cache_etag.erl

```diff
@@ -0,0 +1,174 @@
+    {{strong, Hash}, Req, State};
+
+generate('GetMicroBlockTransactionsCountByHash', Req, State) ->
+    Hash = cowboy_req:binding(hash, Req),
+    {{strong, Hash}, Req, State};
+
+generate('GetGenerationByHash', Req, State) ->
+    handle_generation_hash(Req, State);
+
+generate('GetGenerationByHeight', Req, State) ->
+    handle_generation_height(Req, State);
+
+generate('GetTransactionByHash', Req, State) ->
+    Hash = cowboy_req:binding(hash, Req),
+    {{strong, Hash}, Req, State};
+
+generate('GetTransactionInfoByHash', Req, State) ->
+    Hash = cowboy_req:binding(hash, Req),
+    {{strong, Hash}, Req, State};
+
+generate(_OperationId, Req, State) ->
+    {undefined, Req, State}.
+
+%%%===================================================================
+%%% Private functions
+
+-spec handle_block_height(cowboy_req:req(), any()) -> cowboy_etag_res().
+handle_block_height(Req, State) ->
+    Height0 = cowboy_req:binding(height, Req),
+    case to_int(Height0) of
+        {ok, Height} ->
+            case etag_block_height(Height) of
+                {ok, ETag} ->
+                    {{strong, ETag}, Req, State};
+                undefined ->
+                    {undefined, Req, State}
+            end;
+        _ ->
+            {undefined, Req, State}
+    end.
+
+-spec handle_generation_hash(cowboy_req:req(), any()) -> cowboy_etag_res().
+handle_generation_hash(Req, State) ->
+    Hash0 = cowboy_req:binding(hash, Req),
+    case aeser_api_encoder:safe_decode(key_block_hash, Hash0) of
+        {ok, Hash} ->
+            case etag_generation_hash(Hash) of
+                {ok, ETag} ->
+                    {{strong, ETag}, Req, State};
+                _ ->
+                    {undefined, Req, State}
+            end;
+        _ ->
+            {undefined, Req, State}
+    end.
+
+-spec handle_generation_height(cowboy_req:req(), any()) -> cowboy_etag_res().
+handle_generation_height(Req, State) ->
+    Height0 = cowboy_req:binding(height, Req),
+    case to_int(Height0) of
+        {ok, Height} ->
+            case etag_generation_height(Height) of
+                {ok, ETag} ->
+                    {{strong, ETag}, Req, State};
+                _ ->
+                    {undefined, Req, State}
+            end;
+        _ ->
+            {undefined, Req, State}
+    end.
+
+-spec etag_block_height(integer()) -> {ok, binary()} | undefined.
+etag_block_height(Height) ->
+    case aehttp_logic:get_key_header_by_height(Height) of
+        {ok, Header} ->
+            ETag = etag_block_header(Header),
+            {ok, ETag};
+        _ -> undefined
+    end.
+
+-spec etag_block(aec_blocks:block()) -> binary().
+etag_block(Block) ->
+    Header = aec_blocks:to_header(Block),
+    etag_block_header(Header).
+
+-spec etag_block_header(aec_headers:header()) -> binary().
+etag_block_header(Header) ->
+    {ok, Hash} = aec_headers:hash_header(Header),
+    Hex = aeu_hex:bin_to_hex(Hash),
+    list_to_binary(Hex).
```

Comment:
As the etag is about the representation of the resource - not necessarily only the resource - I wonder if best practice shall include in the etag also e.g. the hash of the swagger.json. (Of course this applies to all endpoints.)

Though it seems good enough.

Reply:
No, this has to be addressed by the API versioning. We already have backward compatibility policy set, it must enforce major version (in the URL as of now v2) bumps if there is backward incompatible changes.

In regards to backward compatible changes, that is new fields only, they will be available once the cache expires.